### PR TITLE
FIX-#3197: do not pass lambdas to the backend in GroupBy

### LINKED ISSUE #3197: Do not pass lambdas to the backend in case of basic GroupBy aggregation
Current GroupBy implementation is still passing python lambdas to the backend to do simple aggregation (for example `mean`):
https://github.com/modin-project/modin/blob/89f6cdde56abf5dcf9561ea6721fb31f84a855d5/modin/pandas/groupby.py#L128-L129
These lambdas are not processable by non-python engines, and so even if the backend does support particular aggregation it can't be executed because it's passed in a python function format. At the moment, all of those lambdas are being passed into a single method `QueryCompiler.groupby_agg`.

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

### Do not pass lambdas to backend

This PR gets rid of passing Python lambdas to the backend as an implementation for basic GroupBy aggregations. Instead of passing lambdas to a single `groupby_agg` method, different aggregations were dedicated into separate query compiler methods. This PR adds 15 new methods to the query compilers API, all of them implemented like:
```python
def groupby_some_aggregation(self, *args, **kwargs):
    return self.groupby_agg(agg_func="some_aggregation", *args, **kwargs)
```
this allows for previous implementations at the different query compilers to work the same as before, because actual computations are still performed by the `groupby_agg`, the only difference is that it now takes string function names instead of callable. It's supposed that if `groupby_agg` does not support passed aggregation it'll default to pandas by calling `groupby_agg` of the ancestor query compiler.

Also, look at the [original issue](https://github.com/modin-project/modin/issues/3197) for discussion about query compilers API bloating and its potential solutions.

### Align backend's GroupBy API
There were some differences in interfaces of different `groupby_*` methods, to not introduce the differ one with the new 15 methods, this PR unifies interface for all of the `groupby_*` methods.

The new interface for every groupby method is the following:
```python
def groupby_method(
    self,
    by,
    axis,
    groupby_kwargs,
    agg_args,
    agg_kwargs
    drop=False,
)
```
Changes to highlight:
1. Removed `is_multi_by` parameter as unusable
2. `QueryCompiler.groupby_agg` now takes "how" parameter, specifying applying strategy of the passed aggregation function: axis-wise (apply the function to each column/row in every group), group-wise (apply the function to every group), or transform-like (apply the function to every group and broadcast the output to the original shape). Previously, `groupby_agg` took a function, that expected a `pandas.DataFrameGroupbyObject` and decided the applying strategy by itself, this approach was pandas-dependent and couldn't be implemented in any other storage formats. 
3. Moved handling of `numeric_only` parameter to the front-end
4. Removed `map_args` and `reduce_args` parameters as they were implementation-specific, replaced them by `agg_args` and `agg_kwargs`, in the case of MapReduce implementation these aggregation parameters are the same for both phases
5. `groupby_args -> groupby_kwargs`

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3197 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/pandas/groupby.py
Comment (dchigarev) [inline modin/pandas/groupby.py:343]: this was copied from #3298, a possible reason for future merge conflicts

File: modin/pandas/groupby.py
Comment (dchigarev) [inline modin/pandas/groupby.py:962]: this will conflict with the same construction in #3298, have to align it with the new behaviour after #3298 will be merged

File: modin/data_management/functions/groupby_function.py
Comment (dchigarev) [inline modin/data_management/functions/groupby_function.py:282]: defaulting to pandas in case of `axis != 0` instead of raising an `AssertionError` (as it was done previously)

File: modin/backends/pandas/query_compiler.py
Comment (dchigarev) [inline modin/backends/pandas/query_compiler.py:2632]: this is a copy of the logic previously presented at the front-end:
https://github.com/modin-project/modin/blob/57bcfc130e9fbe93d41aa6403b80782ef49eb9db/modin/pandas/groupby.py#L286-L289
tests are failing without this condition

File: modin/data_management/functions/default_methods/groupby_default.py
Comment (dchigarev) [inline modin/data_management/functions/default_methods/groupby_default.py:133]: this was added to be able to handle property attributes (like `groupby.dtypes`)

File: modin/utils.py
Comment (dchigarev) [inline modin/utils.py:571]: `functools.partialmethod` does not provide the required result
<details><summary>example</summary>

```python
>>> class Base:
...     def general_method(self, x):
...         print(f"Base: {x}")
...     specific_method_functools = functools.partialmethod(general_method, x=1)
...     specific_method_modinutils = partial_method_late_binding(general_method, x=1)
...
>>> class Derived(Base):
...     def general_method(self, x):
...         print(f"Derived: {x}")
...
>>> # Functools:
>>> Derived().specific_method_functools()
Base: 1
>>> # Desired:
>>> Derived().specific_method_modinutils()
Derived: 1
```

</details>

**Comment by dchigarev** [comment]
Currently, this PR is blocked by #3298 (it also modifies GroupBy API), however current PR is ready for review

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [inline modin/core/dataframe/algebra/default2pandas/groupby.py:126]: Previously, `kwargs["agg_func"]` was holding a function to be applied to a <i>groupby object</i>, now it stores a func to be applied to a <i>Series or a DataFrame</i> representing a certain group (specified via `qc.groupby_agg(how=...)` parameter), that's why we need to replace the direct "agg_func" return to the wrapper.

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [inline modin/core/dataframe/algebra/default2pandas/groupby.py:167]: there are now no branches that return a dictionary in the `cls.get_func`, so removed this if-statement

File: modin/core/dataframe/algebra/groupby.py
Comment (dchigarev) [inline modin/core/dataframe/algebra/groupby.py:286]: fallback to pandas if `axis != 0` instead of throwing an assertion error

File: modin/pandas/groupby.py
Comment (dchigarev) [inline modin/pandas/groupby.py:1046]: the computation of the `.dtypes` property is deferred most of the time, the triggering of its computation would cost us the same [as a regular MapReduce function](https://github.com/modin-project/modin/blob/67013f935861187c784415d4265a5b437dba5690/modin/core/dataframe/pandas/dataframe/dataframe.py#L167-L189) which is kinda sad, maybe we can try to ignore the `numeric_only` parameter at all?

File: modin/backends/base/query_compiler.py
Comment (dchigarev) [reply modin/backends/base/query_compiler.py]: reverted

**Comment by dchigarev** [comment]
@modin-project/modin-core, the PR has been updated and now is ready for review

**Comment by dchigarev** [comment]
@modin-project/modin-core a reminder to review

**Review by YarShev — COMMENTED** [review]
Overall, looks good!

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [reply modin/core/dataframe/algebra/default2pandas/groupby.py:126]: "agg_func" is a parameter of [`qc.groupby_agg`](https://github.com/dchigarev/modin/blob/2d6b03d87c5235f713a3651c0843b3e52047cb07/modin/core/storage_formats/base/query_compiler.py#L2325), we specify it other [groupby qc methods](https://github.com/dchigarev/modin/blob/2d6b03d87c5235f713a3651c0843b3e52047cb07/modin/core/storage_formats/base/query_compiler.py#L2406)

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [reply modin/core/dataframe/algebra/default2pandas/groupby.py:117]: removed `grp` param as unused

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [inline modin/core/dataframe/algebra/default2pandas/groupby.py:98]: removed `grp` parameter as unused

**Comment by dchigarev** [comment]
Windows jobs are red because they [fail to install codecov](https://github.com/modin-project/modin/runs/4806396598?check_suite_focus=true#step:7:26).

**Review by YarShev — APPROVED** [review]

File: modin/pandas/groupby.py
Comment (devin-petersohn) [inline modin/pandas/groupby.py:121]: Are these all just defaulting to pandas now?

File: modin/pandas/groupby.py
Comment (devin-petersohn) [inline modin/pandas/groupby.py:134]: Why is `_check_index` sometimes used and sometimes not?

File: modin/pandas/groupby.py
Comment (dchigarev) [reply modin/pandas/groupby.py:121]: no, all of them do like [`query_compiler.groupby_skew()`](https://github.com/dchigarev/modin/blob/98f54ae176b89556937768cdaafee8ebc06862ab/modin/core/storage_formats/base/query_compiler.py#L2417) -> [`query_compiler.groupby_agg(agg_func="skew")`](https://github.com/dchigarev/modin/blob/98f54ae176b89556937768cdaafee8ebc06862ab/modin/core/storage_formats/base/query_compiler.py#L2428), it's then `storage_format_qc.groupby_agg` responsobility of how to process the function (default to pandas or not), for example, at pandas storage format, `pandas_qc.groupby_agg` always [executes groupby as a full-axis function](https://github.com/modin-project/modin/blob/0faf4675140415e17d4112f9d0d37cfe87770b9e/modin/core/storage_formats/pandas/query_compiler.py#L2785-L2786), so there's no fallback to pandas.

File: modin/pandas/groupby.py
Comment (dchigarev) [reply modin/pandas/groupby.py:134]: `_check_index` and `_check_index_name` do special casing for `as_index=False` when grouping on an index level for certain operations (mimic pandas behavior). These functions were introduced at #3579, the reason why some functions need special handling and others don't is explained in [this comment](https://github.com/modin-project/modin/pull/3579#discussion_r739127976).

At this PR I  just replaced `_apply_agg_func_check_*` with the `_check_*` everywhere in this file (after your comment I spot one place where I missed the `_check_index_name` call, fixed it at the last commit).

File: modin/pandas/groupby.py
Comment (prutskov) [reply modin/pandas/groupby.py:1048]: But we can't ignore this parameter, because it's need for correct computation by partition.groupby internally (see https://github.com/modin-project/modin/pull/3579#discussion_r755105520).

**Review by prutskov — COMMENTED** [review]
Mostly LGTM, left a minor comments

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [reply modin/core/dataframe/algebra/default2pandas/groupby.py:98]: removed FIXME comment as well

File: modin/core/storage_formats/pandas/query_compiler.py
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py]: moved this logic to a separate method [`GroupByDefault.get_aggregation_method`](https://github.com/dchigarev/modin/blob/2c02bada8b4e0376d1653e45a157219c319afddf/modin/core/dataframe/algebra/default2pandas/groupby.py#L555) + added a comment

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [reply modin/core/dataframe/algebra/default2pandas/groupby.py:154]: This function has to mimic `qc.groupby_*` methods interface, so `drop` parameter is required

File: modin/pandas/groupby.py
Comment (vnlitvinov) [reply modin/pandas/groupby.py:1048]: Indeed, let's file a separate tracker (and stop worrying about it in this PR, it's already huge).

File: modin/core/storage_formats/base/query_compiler.py
Comment (vnlitvinov) [inline modin/core/storage_formats/base/query_compiler.py:2404]: I would like to generate those functions in a loop instead of copy-pasting them all over with changing one verb each time, something like
```python
for action, result, agg_func in (('compute the mean value', 'mean value', 'mean'), ('compute unbiased skew', 'unbiased skew', 'skew'), ...):
    @doc_utils.doc_groupby_method(action=action, result=result, refer_to=agg_func)
    def func(self, by, axis, groupby_kwargs, agg_args, agg_kwargs, drop=False):
        return self.groupby_agg(by=by, agg_func=agg_func, axis=axis, groupby_kwargs=groupby_kwargs, agg_args=agg_args, agg_kwargs=agg_kwargs, drop=drop)
    func.__name__ = f'groupby_{agg_func}'
    setattr(GroupByCompiler, func.__name__, func)
```

File: modin/core/storage_formats/pandas/query_compiler.py
Comment (vnlitvinov) [inline modin/core/storage_formats/pandas/query_compiler.py:2693]: why not use a [functools.partial](https://docs.python.org/3.7/library/functools.html#functools.partial)?

File: modin/pandas/groupby.py
Comment (vnlitvinov) [inline modin/pandas/groupby.py:121]: I'm not happy about `type(self._query_compiler).foo`, why do we do this?.. I mean, why `self._query_compiler.foo` is not enough?

File: modin/pandas/groupby.py
Comment (vnlitvinov) [inline modin/pandas/groupby.py:1025]:
```suggestion
        numeric_only : {None, True, False}, default: None
```
plus augment the explanation of the variable? it's no longer a true boolean when it could be any of `True`, `False` and `None`...

File: modin/pandas/groupby.py
Comment (vnlitvinov) [inline modin/pandas/groupby.py:1054]: what's the difference between accessing `self._query_compiler.dtypes` and `self._df.dtypes`?..

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (prutskov) [reply modin/core/dataframe/algebra/default2pandas/groupby.py:154]: Thanks for explanation

**Review by prutskov — APPROVED** [review]
LGTM

File: modin/pandas/groupby.py
Comment (dchigarev) [reply modin/pandas/groupby.py:121]: it's a legacy of how `_wrap_aggregation` was implemented before, it calls the passed method on [modified query compiler](https://github.com/modin-project/modin/blob/34732cbd70381e99bb7c6cbcc90ac60d6f4017aa/modin/pandas/groupby.py#L952-L955), so it shouldn't be bound to the concrete qc.

File: modin/core/storage_formats/base/query_compiler.py
Comment (dchigarev) [reply modin/core/storage_formats/base/query_compiler.py:2404]: `BaseQueryCompiler` is the place where we define QC API, I don't think that any kind of auto-generation can produce a readable interface.

File: modin/core/storage_formats/pandas/query_compiler.py
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py]: used `functools.partial` at the last commit

File: modin/pandas/groupby.py
Comment (dchigarev) [reply modin/pandas/groupby.py]: applied

File: modin/pandas/groupby.py
Comment (dchigarev) [reply modin/pandas/groupby.py]: `self._df` could be a Series, whose `dtypes` property literally stores a `dtype` instead of a list of `dtypes` as DataFrames do. `query_compiler.dtypes` always stores a list of dtypes, that's why it's important to access QC's dtypes rather than `self._df` when iterating on them.

The picked line hasn't caused any problems because it's under `if self._df.ndim == 2` condition, but for the safety I replaced it with the `self._query_compiler.dtypes`.

**Review by prutskov — APPROVED** [review]
LGTM, assuming CI is green

**Review by prutskov — APPROVED** [review]

File: modin/core/storage_formats/base/query_compiler.py
Comment (vnlitvinov) [reply modin/core/storage_formats/base/query_compiler.py:2424]: oh, if that's API definition I agree

**Review by vnlitvinov — APPROVED** [review]
LGTM

**Review by YarShev — APPROVED** [review]

**Pull request merged**
